test/mocha/mock.plugin.js

```diff
@@ -0,0 +1,105 @@
+/*!
+ * Copyright (c) 2017-2018 Digital Bazaar, Inc. All rights reserved.
+ */
+'use strict';
+
+module.exports = {
+  type: 'service',
+  api: {
+    index: async ({createIndexes, collections}) => {
+      should.exist(createIndexes);
+      createIndexes.should.be.a('function');
+      should.exist(collections);
+      collections.should.an('object');
+      should.exist(collections.blockCollection);
+      collections.blockCollection.should.be.a('string');
+      should.exist(collections.eventCollection);
+      collections.eventCollection.should.be.a('string');
+      should.exist(collections.operationCollection);
+      collections.operationCollection.should.be.a('string');
+      return createIndexes([{
+        collection: collections.operationCollection,
+        fields: {'operation.record.type': 1},
+        options: {unique: false, background: false, name: 'fooIndex'}
+      }]);
+    },
+    storage: {
+      operations: {
+        // NOTE: do not use arrow functions here because this function is to be
+        // bound to the class instance. Methods should be properly namespaced
+        // to avoid conflicts.
+        mockQuery: async function({maxBlockHeight, query}) {
```

Comment:
I think this note is weird ... and that this API should appear namespaced under something like `plugins`, declare a unique name for itself, and be passed the storage API so it can use whatever it wants and use whatever function style it wants, etc. Is there a reason not to do it that way?

Reply:
@msporny, thoughts?

Reply:
To be clear, because `bind` is being used to attach the custom methods to the class instance (e.g. events, blocks, operations), one cannot use an anonymous arrow function here.  It *can* however be promise based or callback based.  So custom methods are accessed with:
```
ledgerNode.storage.operations.mockQuery(...);
```
I think your suggestion of adding `plugin` is fine, so this would become:
```
ledgerNode.storage.operations.plugins.mockQuery(...);
```

Now, are you proposing some alternative to this whole scheme?

Reply:
That's pretty close to what I was thinking. I was thinking it would be like this:

```
ledgerNode.storage.plugins.foo.mockQuery(...);
```

I'm not sure plugin functions would only ever use, for example, one API like `operations` or `events` ... they might combine them. So I think we're kind of limited by saying we can only extend those things. It's possible I could be convinced otherwise.

cc: @msporny

Reply:
To be clear again :).  A single plugin may extend as many class instances (e.g. events, blocks, operations.) as it likes.  Does that address your concern?  So, the plugin would just define:
`{operations: {}, events:{}}` etc.

Reply:
Oh, and we've already implemented a plugin that just extends the `operations` class.

Reply:
@mattcollier,

I mean, can it define: `{foo: {}}`? Something top-level that uses both operations and events internally but extends neither directly? How would it be passed the storage API -- what would it be bound to?

Reply:
Just trying to avoid the paint into the corner problem. If we don't support it now but it's clear that we wouldn't have to rewrite how we do plugins to do it in the future, that's good enough for me.

Reply:
IIUC, agreed that we don't want to tie it to the class names we have today (storage, operations, etc.) and personally I'd prefer something like: `{foo: {}}` because we can always do `{foo: { operations {} }}` later... but if we do `{foo: { operations {} }}` now, we're painted into a corner.

Also, I would expect storage plugins to be accessed like so:
`ledgerNode.storage.plugins.foo.mockQuery(...);`

Am I missing something?

Reply:
@msporny can you say more about why you don't want to tie it to the class names we have now?  I think it is logical to `extend` the classes we have now.  We would literally be doing that if it weren't for the fact that we need N plugins to be able to extend the classes and I don't know of a good way to layer that up.  So we're just binding extensions to the class instance.

It feels like some of the concerns being expressed relate to wanting a more generalized approach.  However, storage plugins like this will always be very tightly coupled with the storage method they are associated with `bedrock-ledger-storage-mongodb` in this case.

One of the concepts I had in mind when doing it this way had to do with having some level of control over what plugins had access to.  Such that plugins could (in the future) be restricted to only utilizing one collection or another.  Maybe that's just not a real thing or a valid concern.

Another thought I had was that plugins might actually be able to override core functionality.  Although that seems nice, it also seems very dangerous.